File: src/main/java/com/projectx/university/service/ValidationService.java
package com.projectx.university.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.validation.BeanPropertyBindingResult;
import org.springframework.validation.Errors;
import org.springframework.validation.FieldError;
import org.springframework.validation.ObjectError;
import org.springframework.validation.Validator;
import org.springframework.web.server.ServerWebInputException;

import static java.util.stream.Collectors.joining;

@Service
@RequiredArgsConstructor
public class ValidationService {
    private final Validator validator;

    public <T> T validateInput(final T object) {
        final Errors errors = new BeanPropertyBindingResult(object, object.getClass().getName());
        validator.validate(object, errors);
        if (errors.hasErrors()) {
            final String errorMessages = errors.getAllErrors().stream()
                    .map(ValidationService::getErrorMessage)
                    .sorted()
                    .collect(joining("; "));
            throw new ServerWebInputException(errorMessages);
        }
        return object;
    }

    private static String getErrorMessage(ObjectError error) {
        return error instanceof FieldError
                ? String.format("'%s' property %s", ((FieldError) error).getField(), error.getDefaultMessage())
                : error.getDefaultMessage();
    }
}

File: src/main/kotlin/com/projectx/university/service/ValidationService.kt
package com.projectx.university.service

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import org.springframework.validation.*
import org.springframework.web.server.ServerWebInputException

@Service
class ValidationService(@Qualifier("defaultValidator") private val validator: Validator) {
    fun <T> validate(dto: T, dtoClass: Class<T>): T {
        val errors: Errors = BeanPropertyBindingResult(dto, dtoClass.name)
        validator.validate(dto, errors)
        if (errors.hasErrors()) {
            val errorMessages = errors.allErrors.asSequence()
                    .map { getErrorMessage(it) }
                    .sorted()
                    .joinToString(separator = "; ")
            throw ServerWebInputException(errorMessages)
        }
        return dto
    }

    private fun getErrorMessage(error: ObjectError): String =
            if (error is FieldError)
                "'${error.field}' property ${error.defaultMessage}"
            else error.defaultMessage ?: ""
}